src/ctap2/mod.rs

```diff
@@ -672,6 +660,35 @@ pub fn sign<Dev: FidoDevice>(
             return false;
         }
 
+        // Third, use the shared secret in the extensions, if requested
+        get_assertion.extensions.hmac_secret = match get_assertion
+            .extensions
+            .hmac_secret
+            .take()
+            .map(|hmac_get_secret_or_prf| {
+                if let Some(secret) = dev.get_shared_secret() {
+                    let (extension, selected_credential) = hmac_get_secret_or_prf.calculate(
+                        secret,
+                        &get_assertion.allow_list,
```

Comment:
Should we filter down to an allow-list of length 1 before doing this?

Reply:
We don't know whether we should before we cross-reference `allow_list` with the extension inputs, so the `calculate` method does that internally.

First, the `sign` function here first filters the `allow_list` to just the credentials applicable to the currently chosen authenticator. At this point the `allow_list` should already be reduced to just 1 element in _most_ cases, as it should be rare for an authenticator to have more than one credential for the same account at the same RP. But then second, `calculate` searches for an `evalByCredential` entry matching any item in `allow_list`. If one is found, `calculate` returns that `PublicKeyCredentialDescriptor` and the result handler here sets `allow_list` to contain only that one item on line 676. Does that seem reasonable?

Perhaps the assignment to `get_assertion.allow_list` is currently a bit hidden. Would it be better to try to move that assignment to the top level?

```diff
diff --git a/src/ctap2/mod.rs b/src/ctap2/mod.rs
index 9894ccf..09996ff 100644
--- a/src/ctap2/mod.rs
+++ b/src/ctap2/mod.rs
@@ -661,28 +661,33 @@ pub fn sign<Dev: FidoDevice>(
         }
 
         // Use the shared secret in the extensions, if requested
-        get_assertion.extensions.hmac_secret = match get_assertion
+        (
+            get_assertion.extensions.hmac_secret,
+            get_assertion.allow_list,
+        ) = match get_assertion
             .extensions
             .hmac_secret
             .take()
             .map(|hmac_get_secret_or_prf| {
                 if let Some(secret) = dev.get_shared_secret() {
-                    let (extension, selected_credential) = hmac_get_secret_or_prf.calculate(
+                    hmac_get_secret_or_prf.calculate(
                         secret,
                         &get_assertion.allow_list,
                         pin_uv_auth_result.get_pin_uv_auth_token(),
-                    )?;
-                    if let Some(selected_credential) = selected_credential {
-                        get_assertion.allow_list = vec![selected_credential.clone()];
-                    }
-                    Ok(extension)
+                    )
                 } else {
-                    Ok(hmac_get_secret_or_prf)
+                    Ok((hmac_get_secret_or_prf, None))
                 }
             })
             .transpose()
         {
-            Ok(extension) => extension,
+            Ok(Some((new_hmac_secret, selected_credential))) => (
+                Some(new_hmac_secret),
+                selected_credential
+                    .map(|selected_credential| vec![selected_credential.clone()])
+                    .unwrap_or(get_assertion.allow_list),
+            ),
+            Ok(None) => (None, get_assertion.allow_list),
             Err(e) => {
                 callback.call(Err(e));
                 return false;
```

Reply:
Ok, that sounds reasonable. I must have overlooked the assignment to `get_assertion.allow_list` on my first pass. No need to move it to the top level.